extensions/amp-bind/0.1/bind-expression.js

```diff
@@ -22,25 +23,232 @@ import {parser} from './bind-expr-impl';
 export class BindExpression {
   /**
    * @param {string} expressionString
+   * @throws {Error} On malformed expressions.
    */
   constructor(expressionString) {
     /** @const {string} */
     this.expressionString = expressionString;
+
+    /** {!./bind-expr-defines.ASTNode} */
+    this.ast_ = parser.parse(this.expressionString);
+
+    /** @const {!Object<string, !Object<string, Function>>} */
+    this.functionWhitelist_ = this.createFunctionWhitelist_();
   }
 
   /**
    * Evaluates the expression given a scope.
    * @param {!Object} scope
+   * @throws {Error} On illegal function invocation.
    * @return {*}
    */
   evaluate(scope) {
-    // TODO(choumx): Improve performance by extracting AST construction
-    // and generating evaluator with Function constructor.
-    try {
-      parser.yy = scope;
-      return parser.parse(this.expressionString);
-    } finally {
-      parser.yy = null;
+    return this.eval_(this.ast_, scope);
+  }
+
+  /**
+   * Recursively evaluates and returns value of `node` and its children.
+   * @param {?./bind-expr-defines.ASTNode} node
+   * @param {!Object} scope
+   * @throws {Error}
+   * @return {*}
+   * @private
+   */
+  eval_(node, scope) {
+    if (!node) {
+      return null;
+    }
+
+    const {type, args, value} = node;
+
+    if (type === ASTNodeType.LITERAL) {
+      return value;
+    }
+
+    // Concise helper function for evaluating child nodes.
+    const e = node => this.eval_(node, scope);
+
+    switch (type) {
+      case ASTNodeType.EXPRESSION:
+        return e(args[0]);
+
+      case ASTNodeType.INVOCATION:
```

Comment:
So we can go one step further than this with this: pure, runtime-compiled JS functions. We also keep the security guarantees at compile time (because the lexer only supports our features) and at "evaluate" time (because we can inject our own arbitrary code with the lexed code).


Here's a short example compiling and running an INVOCATION:

```html
<span [text]="text.toLowerCase()" />
```

```js
// parser:
invocation:
  expr '.' NAME args
    %{
      $$.push(`var caller = ${$1}`);
      $$.push(`var method = ${$1}`);
      $$.push(`var callerType = Object.prototype.toString.call(caller)`);
      $$.push(`var whitelist = SOME_GLOBAL[callerType]`);
      $$.push(`if (whitelist) {`);
      $$.push(`  var func = caller[method];`);
      $$.push(`  if (func && func === whitelist[method]) {`);
      $$.push(`    if (containsObject_(params)) {`);
      $$.push(`      throw new Error('Unexpected argument type in ' + method ' + }().');`);
      $$.push(`    } else {`);
      $$.push(`      caller[method](${$4})`);
      $$.push(`    }`);
      $$.push(`  }`);
      $$.push(`}`);
      $$.push(`throw new Error(method + '() is not a supported function.');`);
    %}
  ;

//parsing
this.fn_ = new Function('scope', parser.parse(this.expressionString));

//eval
this.fn_(scope);